File: leetcode/subsets.js
//Leetcode 78: Subsets
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
  let temp = [];
  let result = [];

  function helperFun(nums, i, temp) {
    if (i === nums.length) {
      result.push(JSON.parse(JSON.stringify(temp)));
      return;
    }
    temp.push(nums[i]);
    helperFun(nums, i + 1, temp);
    temp.pop(nums[i]);
    helperFun(nums, i + 1, temp);
  }
  helperFun(nums, 0, temp);
  return result;
};

let nums = [1, 2, 3];
console.log(subsets(nums));

File: leetcode/subsets.py
# Leetcode 78. Subsets
from typing import List

class Solution:
  def subsets(self, nums: List[int]) -> int:
    n = len(nums)
    res, sol = [], []

    def backtrack(i):
      if i == n:
        res.append(sol[:])
        return
      
      sol.append(nums[i])
      backtrack(i + 1)
      sol.pop()

      backtrack(i + 1)
    
    backtrack(0)
    return res


if __name__ == "__main__":
  sol = Solution()
  print(sol.subsets([1, 2, 3]))

